modules/developers-guide/pages/tutorials/reproducible-builds.adoc

```diff
@@ -0,0 +1,341 @@
+     transitive dependencies of your project that satisfy the requirements specified in your `package.json`.
+     However, `npm install` will overwrite the `package-lock.json` file every time it is invoked.
+     Thus, once you are ready to create the final version of your canister, run `npm install` only once.
+     After that, commit `package-lock.json` to your version control system.
+     Finally, when checking the build for reproducibility, use `npm ci` instead of `npm install`.
+
+   * For Rust code, Cargo will automatically generate a `Cargo.lock` file with the fixed versions of
+     your (transitive) dependencies.
+     Like with `package-lock.json`, you should commit this file to your version control system once you
+     are ready to produce the final version of your canister.
+     Furthermore, Cargo by default ignores the locked versions of dependencies.
+     Pass the `--locked` flag to the `cargo` command to ensure that the locked dependencies are used.
+
+   * You have to allocate canister IDs in advance, as canisters refer to each other by their IDs.
+
+.  Your own build scripts must not introduce non-determinism.
+   Obvious sources of non-determinism include randomness, timestamps, or code obfuscators.
+   Less obvious sources include locales, absolute file paths, order of files in a directory, and remote URLs whose content can change.
+   Furthermore, relying on third-party build plug-ins exposes you to any non-determinism introduced by these.
+
+.  Given the same dependencies and deterministic build scripts, the build tools themselves (`moc` for Motoko, `cargo` for Rust, `webpack` 
+   by default for front-end development) must also be deterministic.
+   The good news is that all of these tools aim to be deterministic.
+   However, they are complicated pieces of software, and ensuring determinism is non-trivial.
+   Thus, non-determinism bugs can and do occur.
+   For Rust, see the https://github.com/rust-lang/rust/labels/A-reproducibility[list of current potential non-determinism issues in Rust].
+   Furthermore, we have observed differences between Rust code compiled to WASM under Linux and MacOS, and thus recommend pinning the build platform and its version.
+   For webpack, determinism is important to ensure caching, and webpack has since version 5 introduced https://webpack.js.org/configuration/optimization/[deterministic naming of module and chunk IDs] that you should use.
+   The Motoko compiler aims to be deterministic, but there are currently known issues.
+   Absolute file paths might be included in the binary, and the Motoko runtime bundled in the canisters may differ between platforms (such as Linux and MacOS).
+   You can work around these issues by performing your build in a Docker container, as this will fix both the platform and the file paths.
+   We will try to address these issues in the future.
+   If you find further reproducibility issues with `moc`, we would like to hear about them.
```

Comment:
Link to desired reporting of non-determinism issues?

Reply:
I'll add a link to https://github.com/dfinity/motoko/issues - @crusso , please yell if this is the wrong place